File: src/precedence.js
//*-- Precedence
/**
 * Precedence defines the order in wihch tokens are
 * treated by the parser. 
 */
function Precedence(/*String*/ name,
                    /*int*/ ordinal) /*implements Comparable<Precedence>*/ {
    this._name = name;
    this._ordinal = ordinal;
}

/**
 * Returns the representative name of this Precedence. 
 */
Precedence.prototype.getName = /*String*/ function() {
    return this._name;
}

/**
 * Returns the ordinal of this Precedence. 
 */
Precedence.prototype.getOrdinal = /*int*/ function() {
    return this._ordinal;
}

/**
 * Compares two Precedence.
 */
Precedence.prototype.compareTo = /*int*/ function(/*Precedence*/ other) {
    return this._ordinal - other._ordinal;
};

// Predefined precedences
/*Precedence*/ Precedence.Expression = new Precedence("expression", -100);
/*Precedence*/ Precedence.Zero = new Precedence("zero", 0);
/*Precedence*/ Precedence.Assign = new Precedence("assign", 1);
/*Precedence*/ Precedence.LogicalOr = new Precedence("logicalOr", 2);
/*Precedence*/ Precedence.LogicalAnd = new Precedence("logicalAnd", 3);
/*Precedence*/ Precedence.BitwiseOr = new Precedence("bitwiseOr", 4);
/*Precedence*/ Precedence.BitwiseAnd = new Precedence("bitwiseAnd", 5);
/*Precedence*/ Precedence.BitwiseXor = new Precedence("bitwiseXor", 6);
/*Precedence*/ Precedence.Equality = new Precedence("equality", 7);
/*Precedence*/ Precedence.Relational = new Precedence("relational", 8);
/*Precedence*/ Precedence.Shift = new Precedence("shift", 9);
/*Precedence*/ Precedence.Additive = new Precedence("additive", 10);
/*Precedence*/ Precedence.Multiplicative = new Precedence("multiplicative", 11);

// export
module.exports = Precedence;

File: lib/src/language/common/ast/precedence.dart
library mammouth.language.common.ast.precedence;

//*-- Precedence
/**
 * Defines the priority in which tokens are treated by parser.
 */
class Precedence implements Comparable<Precedence> {
  const Precedence(this.ordinal);

  /**
   * The ordinal of the precedence, tokens with precedence of higher ordinal
   * have more priority.
   */
  final int ordinal;

  /// Compares this precedence with given precedence [other].
  @override
  int compareTo(Precedence other) {
    return this.ordinal - other.ordinal;
  }

  bool operator <(Precedence other) {
    return this.compareTo(other) < 0;
  }

  bool operator <=(Precedence other) {
    int comp = this.compareTo(other);
    return comp < 0 || comp == 0;
  }

  bool operator >(Precedence other) {
    return this.compareTo(other) > 0;
  }

  bool operator >=(Precedence other) {
    int comp = this.compareTo(other);
    return comp > 1 || comp == 0;
  }

  Precedence operator +(int i) {
    return Precedence.values[
    (Precedence.values.indexOf(this) + i) % Precedence.values.length];
  }

  static const Precedence Zero = const Precedence(0); // default
  static const Precedence Assign = const Precedence(
      1); // "=", "+=", "-=", "*=", "%="
  static const Precedence LogicalOr = const Precedence(2); // "||"
  static const Precedence LogicalAnd = const Precedence(3); // "&&"
  static const Precedence BitwiseOr = const Precedence(4); // "|"
  static const Precedence BitwiseAnd = const Precedence(5); // "&"
  static const Precedence BitwiseXor = const Precedence(6); // "^"
  static const Precedence Equality = const Precedence(7); // "==", "!="
  static const Precedence Relational = const Precedence(
      8); // "<", ">", "<=", ">="
  static const Precedence Shift = const Precedence(9); // "<<", ">>"
  static const Precedence Additive = const Precedence(10); // "+", "-"
  static const Precedence Multiplicative = const Precedence(
      11); // "*", "**", "/", "%"

  /// List of all possible precedence values.
  static const List<Precedence> values = const <Precedence>[
    Zero,
    Assign,
    LogicalOr,
    LogicalAnd,
    BitwiseOr,
    BitwiseAnd,
    BitwiseXor,
    Equality,
    Relational,
    Shift,
    Additive,
    Multiplicative
  ];
}
